File: concrete/src/main/java/pcb/uwu/amount/finance/USD.java
package pcb.uwu.amount.finance;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.finance.CurrencyUnit;
import pcb.uwu.utils.UnitAmountUtils;

import java.math.BigDecimal;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class USD extends Money {

	public static final CurrencyUnit CURRENCY = CurrencyUnit.of("USD");

	// region constructor

	public USD(Number value) {
		super(value, CURRENCY);
	}

	public USD(String value) {
		super(value, CURRENCY);
	}

	public USD(BigDecimal value) {
		super(value, CURRENCY);
	}

	public USD(BigDecimalAmount amount) {
		super(amount, CURRENCY);
	}

	// endregion

	// region implement UnitAmount

	public USD plus(USD other) {
		return new USD(getAmount().plus(other.getAmount().getValue()));
	}

	public USD minus(USD other) {
		return new USD(getAmount().minus(other.getAmount().getValue()));
	}

	@NotNull
	@Override
	public USD plus(@NotNull UnitAmount<CurrencyUnit> other) {
		return new USD(plusAmount(this, other));
	}

	@NotNull
	@Override
	public USD minus(@NotNull UnitAmount<CurrencyUnit> other) {
		return new USD(minusAmount(this, other));
	}

	@NotNull
	@Override
	public USD times(@NotNull Number other) {
		return new USD(UnitAmountUtils.times(this, other));
	}

	@NotNull
	@Override
	public USD div(@NotNull Number other) {
		return new USD(dividedByScalar(this, other));
	}

	// endregion
}

File: concrete/src/main/java/pcb/uwu/amount/base/currency/USD.kt
package pcb.uwu.amount.base.currency

import pcb.uwu.amount.base.Currency
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.CurrencyUnit
import pcb.uwu.unit.base.CurrencyUnit.Companion.USD

class USD : Currency
{
    @JvmOverloads
    constructor(amount: Number, magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, USD)

    @JvmOverloads
    constructor(value: String, magnitude: Magnitude = NATURAL)
            : super(value, magnitude, USD)

    // region UnitAmount

    override fun plus(currency: UnitAmount<CurrencyUnit>) =
        USD(this.amount + (currency to USD).amount)

    override fun minus(currency: UnitAmount<CurrencyUnit>) =
        USD(this.amount - (currency to USD).amount)

    override fun times(scalar: Number) =
        USD(this.amount * scalar)

    override fun div(scalar: Number) =
        USD(this.amount / scalar)

    // endregion
}